### PR TITLE
[Validator] Add support of nested attributes

### LINKED ISSUE #38503: [Validator] Use composite constraints as attributes
On our path towards configuring validation with PHP attributes (#38096), after #38309, #38382, #38410 and #38499, composite constraints are still open. I'd like to work on them for the 5.3 release and therefore continue the discussion that we started on #38309.

As a reference, those are the constraints that we're talking about.

* `All`
* `AtLeastOneOf`
* `Collection`
* `Compound` (abstract)
* `Existence` (abstract)
    * `Required`
    * `Optional`
* `Sequentially`

All those constraints need a set of nested constraints as input. As discussed earlier, those are a bit tricky because PHP 8.0 does not allow us to nest attributes. If we want to use attributes here as well, we need to work around that limitation.

While most of the constraints receive the nested constraints as simple array, `Collection` however requires a mapping (field to constraint) and is usually combined with other composite constraints, which gives us a second nesting level.

Let's discuss the options that we have.

### Pray and hope for PHP 8.1 to deliver nested attributes

That's the easiest one for us! Let's do an RFC.

### Introduce an end marker and flatten the nesting

Suggested by @nicolas-grekas (https://github.com/symfony/symfony/pull/38309#issuecomment-699992236):

```php
#[Assert\All,
    Assert\NotNull,
    Assert\Length(max: 6),
EndAssert]
```

In this case, our `AnnotationLoader` would traverse the constraints following a composite until it finds the `EndAssert` attribute. Those attributes would then be funnelled into the property labelled as `Composite::getCompositeOption()`. The only case where is approach won't work out of the box would be `Collection` because we would just generate an ordered list this way, not a mapping like `Collection` requires.

### Allow the list of constraints to be passed as callable

```php
#[Assert\All(
    constraints: [MyConstraintFactory::class, 'createConstraints'],
)]
```

While constructing the constraint, the callable would be resolved.

This would be an easy workaround, but of course it defeats the main reason one would want to use attributes: explicit readable configuration at the class/method/property that is to be configured.

### Introduce an array convention for nested constraints

```php
#[Assert\All(
    constraints: [
        [Assert\NotNull::class],
        [Assert\Length::class, 'max' => 6],
    ]
)]
```

When constructing the constraint, each of the arrays would be replaced by the corresponding constructed constraint. This would be easily doable, however we would lose IDE support for the nested constraint declarations.

### Document workarounds

A possible workaround would be building an own constraint that extends and configures a composite constraint. We can document how to do this. If we expect nested attributes to be implemented eventually, those workarounds could be enough for the time being.

### PR DESCRIPTION
**Pull request by alexandre-daubois**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #38503
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/15541

Although the RFC (https://wiki.php.net/rfc/new_in_initializers) is in the voting phase (until 14 July), it is already well on its way to passing.

Based on @nikic's development (https://github.com/php/php-src/pull/7153), this makes the development of support possible. It will obviously take a little while before this pull request is merged.

If this pull request is OK for you, I'll get to work on writing the existing documentation for the attribute validation constraints.

Capture d’écran du 2021-07-05 17-11-23

Not sure about the Symfony version to target, as `AtLeastOneOf` has been introduced in 5.1. Although, I couldn't find attributes validation documentation for 4.4.

### CODE REVIEW

**Comment by OskarStark** [comment]
As this is a new feature it should target `5.4` branch

**Comment by alexandre-daubois** [comment]
Alright, I'll rebase to 5.4!

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/Validator/Tests/Fixtures/Attribute/Nested/Entity.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Tests/Fixtures/Attribute/Nested/Entity.php]: the Ci reports `PHP Fatal error:  Constant expression contains invalid operations`

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
Let's wait for the RFC to be merged before merging :)

File: src/Symfony/Component/Validator/Tests/Fixtures/Attribute/Nested/Entity.php
Comment (derrabus) [inline src/Symfony/Component/Validator/Tests/Fixtures/Attribute/Nested/Entity.php]:
```suggestion
        Assert\All(
            constraints: [
                new Assert\NotNull(),
                new Assert\Range(min: 3),
            ],
        ),
```

Let's switch to named arguments, like we did for the other constraints.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Validator/Tests/Mapping/Loader/AnnotationLoaderTest.php
Comment (derrabus) [inline src/Symfony/Component/Validator/Tests/Mapping/Loader/AnnotationLoaderTest.php]:
```suggestion
            yield 'nested_attributes' => ['Symfony\Component\Validator\Tests\Fixtures\NestedAttribute'];
```

Let's keep the directory structure flat

File: src/Symfony/Component/Validator/Tests/Fixtures/Attribute/Nested/Entity.php
Comment (alexandre-daubois) [reply src/Symfony/Component/Validator/Tests/Fixtures/Attribute/Nested/Entity.php:40]: Looking into it!

File: src/Symfony/Component/Validator/Tests/Mapping/Loader/AnnotationLoaderTest.php
Comment (alexandre-daubois) [reply src/Symfony/Component/Validator/Tests/Mapping/Loader/AnnotationLoaderTest.php:190]: Alright!

File: src/Symfony/Component/Validator/Tests/Fixtures/Attribute/Nested/Entity.php
Comment (alexandre-daubois) [reply src/Symfony/Component/Validator/Tests/Fixtures/Attribute/Nested/Entity.php]: I set `mixed` as type of constructor parameter, because of this:

https://github.com/symfony/symfony/blob/e34cd7dd2c6d0b30d24cad443b8f964daa841d71/src/Symfony/Component/Validator/Tests/Constraints/AllValidatorTest.php#L30

Comment (alexandre-daubois) [reply src/Symfony/Component/Validator/Tests/Fixtures/Attribute/Nested/Entity.php]: Didn't type-hinted the constructor parameter because of this:

https://github.com/symfony/symfony/blob/e34cd7dd2c6d0b30d24cad443b8f964daa841d71/src/Symfony/Component/Validator/Tests/Constraints/AllValidatorTest.php#L30

**Comment by alexandre-daubois** [comment]
Added missing tests for `Sequentially` and `AtLeastOneOf` attributes.
Added support for `Collection` as well (https://github.com/symfony/symfony/pull/41994/files#diff-d9f229d35c24533de28e1e630ee4e74afa4ce27d7af8f137d788898e0b9efc73R41).
Added documentation pull-request.

`Required` and `Optional` don't need update. I still added them in AnnotationLeaderTest.

File: src/Symfony/Component/Validator/Tests/Fixtures/Attribute/Nested/Entity.php
Comment (alexandre-daubois) [reply src/Symfony/Component/Validator/Tests/Fixtures/Attribute/Nested/Entity.php]: Done!

**Comment by derrabus** [comment]
In order to get the PHP 8.0 testsuite green, you need to exclude your new fixtures in `.github/patch-types.php`.

**Review by derrabus — APPROVED** [review]
LGTM, once patch-types.php has been updated.

File: src/Symfony/Component/Validator/Tests/Mapping/Loader/AnnotationLoaderTest.php
Comment (derrabus) [inline src/Symfony/Component/Validator/Tests/Mapping/Loader/AnnotationLoaderTest.php]:
```suggestion
```

Unused import.

**Comment by alexandre-daubois** [comment]
> In order to get the PHP 8.0 testsuite green, you need to exclude your new fixtures in `.github/patch-types.php`.

Will do, thanks!

**Review by derrabus — APPROVED** [review]

**Comment by fabpot** [comment]
This is now available in the 8.1 PHP branch.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @alexandre-daubois.

**Comment by chekalsky** [comment]
@alexandre-daubois @fabpot `Optional` is still not supported in Attributes if I am not mistaken, is it done consciously? Haven't found any mentions of it.

**Comment by xabbuh** [comment]
@chekalsky Did you experience any issues when using `Optional` or `Required` inside an attribute? As far as I can see this should work (and we also have a test case for this).

File: src/Symfony/Component/Validator/Tests/Fixtures/NestedAttribute/Entity.php
Comment (xabbuh) [inline src/Symfony/Component/Validator/Tests/Fixtures/NestedAttribute/Entity.php:49]: @chekalsky this is the test I mean

**Comment by chekalsky** [comment]
@xabbuh Yes, but only when it is not nested (e.g. I am trying to make one of the properties itself optional). Does it make sense?

**Comment by stof** [comment]
Well, `Optional` and `Required` don't make sense as top-level attributes. They are meant to be used inside the `Collection` constraint.

**Pull request merged**
